pkg/alertmanager/multitenant.go

```diff
@@ -737,6 +766,59 @@ func (am *MultitenantAlertmanager) ServeHTTP(w http.ResponseWriter, req *http.Re
 		return
 	}
 
+	if am.cfg.ShardingEnabled && am.distributor.IsPathSupported(req.URL.Path) {
+		am.distributor.DistributeRequest(w, req)
+		return
+	}
+
+	// If sharding is not enabled or Distributor does not support this path,
+	// it is a simple pass-through.
+	am.serveHTTP(w, req)
+}
+
+// HandleRequest serves the Alertmanager's web UI and API sent via gRPC.
+func (am *MultitenantAlertmanager) HandleRequest(ctx context.Context, in *alertmanagerpb.Request) (*alertmanagerpb.Response, error) {
```

Comment:
nit: I have the feeling we can use [this handler](https://github.com/weaveworks/common/blob/master/httpgrpc/server/server.go#L50-L71) to reduce some of the boilerplate but I guess we could do that in a follow up PR.

Reply:
A future PR sounds good. At the moment I am sending all non 200 as errors so that the quorum can catch the failures of any kind properly. The linked handler would return errors only for 5xx and not 4xx hence will require some special handling in the distributor.

Reply:
After some new discussions, I have used this ready handler now and modified distributor code accordingly.